registry/distribution/registry.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+// Package distribution provides a client to the remote registry.
+// Reference: https://github.com/distribution/distribution
+package distribution
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"oras.land/oras-go/v2/registry"
+)
+
+// RepositoryOptions is an alias of Repository to avoid name conflicts.
+// It also hides all methods associated with Repository.
+type RepositoryOptions Repository
+
+// Registry is a HTTP client to a remote registry.
+type Registry struct {
+	// RepositoryOptions contains common options for Registry and Repository.
+	// It is also used as a template for derived repositories.
+	RepositoryOptions
+
+	// RepositoryListPageSize specifies the page size when invoking the catalog
+	// API.
+	// If zero, the page size is determined by the remote registry.
+	// Reference: https://docs.docker.com/registry/spec/api/#catalog
+	RepositoryListPageSize int
```

Comment:
As mentioned in the documentation, the value can be overrided per request at the API call level. I will improve it to per request the HTTP call level.

Be desgin, you can leave it zero for convenience. It seems dangerous at the first galance since the server may flood the client. However, the max size of the server response is limited by `MaxMetadataBytes`, and thus the max `RepositoryListPageSize` is implicitly limited.